File: src/GUI/objects/Table/TableInfo.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package GUI.objects.Table;

import static java.util.Collections.EMPTY_LIST;
import java.util.List;
import java.util.function.BiFunction;
import javafx.beans.Observable;
import javafx.scene.control.Labeled;
import javafx.scene.control.TableView;
import static org.atteo.evo.inflector.English.plural;
import util.TODO;
import static util.TODO.Purpose.BUG;
import static util.TODO.Severity.LOW;
import static util.Util.copySelectedItems;

/**
 * Provides information about table items and table item selection.
 * 
 * @param <E> type of table element
 * 
 * @author Plutonium_
 */
public final class TableInfo<E> {
    
    /** Default text factory. Provides texts like: <pre>
     * 'All: 1 item'
     * 'Selected: 89 items'
     * </pre>
     * Custom implementation change or expand the text with additional 
     * information depending on type of table elements. */
    public static final BiFunction<Boolean,List<?>,String> DEFAULT_TEXT_FACTORY = (all, list) -> {
        String prefix1 = all ? "All: " : "Selected: ";
        int s = list.size();
        return prefix1 + s + " " + plural("item", s);
    };
    
    /** The graphical text element */
    public Labeled node;
    /** Provides text to the node. The first parameters specifies whether selection
     * is empty, the other is the list of table items if selection is empty or 
     * selected items if nonempty. */
    public BiFunction<Boolean,List<E>,String> textFactory = (BiFunction)DEFAULT_TEXT_FACTORY;
    
    /** Sets the node. The text updating needs to be done manually. */
    public TableInfo(Labeled node) {
        this.node = node;
    }
    
    @TODO(purpose = BUG, severity = LOW, note = "see the commented part"
            + "modification erros - INVSTIGATE if te issue still appears (some sort of IndexOutOfBounds in Table SortedList)")
    /** Sets the node and listeners to update the text automatically by monitoring
      * the table items and selection. */
    public TableInfo(Labeled node, TableView<E> t) {
        this(node);
        List<E> items = t.getItems();
        // the below causes bug & not sure if EMPTY_LIST does not produce incorrect
        // behavior, but looks fine so far the problem seems to be when the
        // whole table list changes and both of the
        // listeners are called, just a guess
        // t.getItems().addListener((Observable o)-> updateText(items, copySelectedItems(t)));
        t.getItems().addListener((Observable o)-> updateText(items, EMPTY_LIST));
        t.getSelectionModel().getSelectedItems().addListener((Observable o)-> updateText(items, copySelectedItems(t)));
        updateText(items, EMPTY_LIST);
    }
    
    /** Updates the text of the node using the text factory.
     * @param all all items of the table
     * @param selected  selected items of the table */
    public final void updateText(List<E> all, List<E> selected) {
        boolean isAll = selected.isEmpty();
        List<E> l = isAll ? all : selected;
        node.setText(textFactory.apply(isAll,l));
        
        // if bugs appear avoid using original list by copying it
        node.setText(textFactory.apply(isAll,l));
    }
}

File: src/player/main/sp/it/pl/gui/nodeinfo/TableInfo.kt
package sp.it.pl.gui.nodeinfo

import javafx.scene.control.Labeled
import javafx.scene.control.TableView
import sp.it.util.reactive.Subscription
import sp.it.util.reactive.onChange
import sp.it.util.reactive.plus
import sp.it.util.reactive.syncNonNullWhile
import sp.it.util.text.pluralUnit
import kotlin.properties.Delegates.observable

/**
 * Provides information about table items and table item selection.
 *
 * @param <E> type of table element
 */
class TableInfo<E>: NodeInfo<TableView<E>> {

    private var updateText = {}
    private var listMonitorDisposer: Subscription? = null
    /** The graphical text element */
    val node: Labeled
    /**
     * Provides text to the node. The first parameters specifies whether selection
     * is empty, the other is the list of table items if selection is empty or
     * selected items if nonempty.
     */
    var textFactory by observable<(Boolean, List<E>) -> String>(DEFAULT_TEXT_FACTORY) { _, _, _ -> updateText() }

    constructor(node: Labeled) {
        this.node = node
    }

    /**
     * Creates infoTable for the labeled and invokes [bind] with the table
     */
    constructor(node: Labeled, t: TableView<E>): this(node) {
        bind(t)
    }

    override fun setVisible(v: Boolean) {
        node.isVisible = v
    }

    override fun bind(bindable: TableView<E>) {
        unbind()
        listMonitorDisposer = bindable.itemsProperty().syncNonNullWhile {
            val listAll = it
            val listSelected = bindable.selectionModel.selectedItems
            updateText = { updateText(bindable.items, listSelected) }

            updateText()
            val s1 = listAll.onChange(updateText)
            val s2 = listSelected.onChange(updateText)
            s1+s2
        }
    }

    override fun unbind() {
        listMonitorDisposer?.unsubscribe()
        updateText = {}
    }

    /**
     * Updates the text of the node using the text factory.
     *
     * @param allItems all items of the table
     * @param selectedItems selected items of the table
     */
    fun updateText(allItems: List<E>, selectedItems: List<E>) {
        val isAll = selectedItems.isEmpty()
        val items = if (isAll) allItems else selectedItems
        node.text = textFactory(isAll, items)
    }

    companion object {

        /** Default text factory. Provides texts like 'All: 1 item' or 'Selected: 89 items'. */
        val DEFAULT_TEXT_FACTORY: (Boolean, List<*>) -> String = { isSelectionEmpty, list ->
            val prefix = if (isSelectionEmpty) "All: " else "Selected: "
            val size = list.size
            "$prefix "+"item".pluralUnit(size)
        }

    }
}